### ISSUE TITLE
[M1] What about the support for Apple M1 processors?

### ISSUE THREAD

**Issue by ebellumat** [yuzu-emu/yuzu#6184]
Both Nintendo Switch and Apple M1 processors use ARM on the architecture, i think to add the support for these machines would be great.

**Comment by renA21** [comment]
Some work has been done to run yuzu on M1-based Macs, however, the problem is that it's GPU drivers are only available in macOS and only support the Metal graphical backend. The Metal API is currently fairly limited that its unable to fully emulate functions for rendering Switch graphics.

**Comment by ebellumat** [comment]
Understood. I never compared the Metal with open graphics API, i was think that'll be equivalent.

**Comment by FernandoS27** [comment]
> Understood. I never compared the Metal with open graphics API, i was think that'll be equivalent.

nope, Apple didn't add many common features used by other APIs. Many of those vital for switch emulation.

**Comment by iMonZ** [comment]
> > Understood. I never compared the Metal with open graphics API, i was think that'll be equivalent.
> 
> nope, Apple didn't add many common features used by other APIs. Many of those vital for switch emulation.

Actually that’s not true.
We have MoltenVK and it works fine enough

**Comment by ebellumat** [comment]
I'm reopening this thread, Metal 3 will be launched in the new Mac OS soon, they have all the features missing on the new APi's, such as mesh shaders, ray-tracing, their own "DLSS", we can also use this to improve emulation performance on the GPU. 

We just need to wait to MoltenVK to update or go directly into Metal rendering in Bare Metal. 

https://developer.apple.com/metal/

**Comment by ebellumat** [comment]
So, we have two options to start the porting.

1. Via Rosetta -  This way we can implement Metal and check if works before translates to ARM64.
2. Port Yuzu to ARM64 Linux befores goes to the Mac OS version for M1.

**Comment by liamwhite** [comment]
System emulation requires a patch to dynarmic due to an unfixed bug in Rosetta: https://github.com/merryhime/dynarmic/commit/c60fd3f0ac8165c743ff9c11c300a1a99089fc7c 
Dynarmic's ARM64 version is currently in progress and is the only real obstacle to system porting.

Graphics emulation should possible today with a software renderer like Lavapipe, which I have successfully used on a system without a graphics card. Hardware support still require additional features that Metal does not support (as per https://developer.apple.com/documentation/metal, the latest version):
- 8 bit vertex and index formats
- Line loops
- Triangle fans
- Quads/quad strips
- Polygon topology
- Patches topology
- Defined aliasing between different memory types
- Native clip control ([-1,1] vs [0,1])
- Masked color clears
- Communication channels to report driver issues

**Comment by ericcurtin** [comment]
For the record I think the ARM64 Linux port would be most ideal here... Development could even be started on a Raspberry Pi 3/4... I'm working with a team of developers and we will have a version of Fedora with GPU accelerated graphics running natively in 2023 (and a version of native Fedora without accelerated graphics before the end of 2022)... Follow Asahi Lina/Hector Martin/Alyssa Rosenzweig/Asahi Linux on Twitter or YouTube for the low level details https://twitter.com/LinaAsahi?t=G_pnEfMIVEnaYk0_2dBvUw&s=09

**Comment by inzar98** [comment]
> System emulation requires a patch to dynarmic due to an unfixed bug in Rosetta: [merryhime/dynarmic@c60fd3f](https://github.com/merryhime/dynarmic/commit/c60fd3f0ac8165c743ff9c11c300a1a99089fc7c) Dynarmic's ARM64 version is currently in progress and is the only real obstacle to system porting.
> 
> Graphics emulation should possible today with a software renderer like Lavapipe, which I have successfully used on a system without a graphics card. Hardware support still require additional features that Metal does not support (as per https://developer.apple.com/documentation/metal, the latest version):
> 
> * 8 bit vertex and index formats
> * Line loops
> * Triangle fans
> * Quads/quad strips
> * Polygon topology
> * Patches topology
> * Defined aliasing between different memory types
> * Native clip control ([-1,1] vs [0,1])
> * Masked color clears
> * Communication channels to report driver issues

Any news with metal 3 ?

**Comment by liamwhite** [comment]
There are no relevant updates. Support for most features we need is still missing, and only compilation and support for homebrew demos is feasible currently.

**Comment by liamwhite** [comment]
Also, it is possibly worth pointing out that we are able to run games and such on Asahi Linux using Lavapipe, a software renderer, but it only compiles on Linux due to integration with the DRM subsystem.

![](https://cdn.discordapp.com/attachments/402002111241388032/1042580958047645807/Screenshot_20221116_182106.png)

**Comment by inzar98** [comment]
> Also, it is possibly worth pointing out that we are able to run games and such on Asahi Linux using Lavapipe, a software renderer, but it only compiles on Linux due to integration with the DRM subsystem.
> 
> ![](https://camo.githubusercontent.com/f6fd12f52d9ac89a606767667aa5ebd65f56f02c0ef917beb5f523cf2385ad75/68747470733a2f2f63646e2e646973636f72646170702e636f6d2f6174746163686d656e74732f3430323030323131313234313338383033322f313034323538303935383034373634353830372f53637265656e73686f745f32303232313131365f3138323130362e706e67)

oh cool. So we only need gpu driver for asahi right ?

**Comment by inzar98** [comment]
Well [Ryujinx somehow ported to macOS](https://blog.ryujinx.org/the-impossible-port-macos/) I prefer yuzu over ryujinx so can we expect any new soon ?

**Comment by iMonZ** [comment]
> 

I say 2
